### PR TITLE
Add .NET Core support

### PR DESCRIPTION
**Pull request by am11**
This is a sequel of maxmind/MaxMind-DB-Reader-dotnet#27.

Notes:

### How to build:

* Install Latest dotnet/cli via installer: https://github.com/dotnet/cli/#installers-and-binaries

 Or

* via command-line (standalone dotnet runtime+msbuild+nuget all in one portable directory):

  Windows:

  ```powershell
  # this is PowerShell

  cd GeoIP2-dotnet

  mkdir scripts

  wget https://raw.githubusercontent.com/dotnet/cli/rel/1.0.0/scripts/obtain/install.ps1 -OutFile .\scripts\install.ps1

  $DOTNET_INSTALL_DIR = "$pwd\.dotnetcli" 

  .\scripts\install.ps1 -Channel preview -version Latest -InstallDir "$DOTNET_INSTALL_DIR" -NoPath

  $env:Path = "$DOTNET_INSTALL_DIR;$env:Path"

  # test --info (or --version)
  dotnet --info
  ```

  Unix:

  ```bash
  # this is Unix Shell

  cd GeoIP2-dotnet

  curl -L --create-dirs https://raw.githubusercontent.com/dotnet/cli/rel/1.0.0/scripts/obtain/install.sh -o ./scripts/install.sh

  export DOTNET_INSTALL_DIR="$PWD/.dotnetcli"

  bash ./scripts/install.sh --channel "preview" --version Latest --install-dir "$DOTNET_INSTALL_DIR" --no-path

  export PATH="$DOTNET_INSTALL_DIR:$PATH"

  # test --info (or --version)
  dotnet --info
  ```

* Restore + Build

  ```sh
  # OS-agnostics build commands
  # in GeoIP2-dotnet directory

  dotnet restore

  cd GeoIP2.UnitTests

  dotnet run # this will build dependency (GeoIP2), then build the UnitTests project,
             # executes the test runner and show results.
  ```

Note that [`RichardSzalay.MockHttp`](https://github.com/richardszalay/mockhttp) dependency in `UnitTests` does not currently supports NetStandard, so I have `#if`'d out the test class which is using those mocks for `NETCOREAPP1.0`.

I also made some minor code improvements along the way.

### CODE REVIEW

**Comment by am11** [comment]
@oschwald, please take a look.

In this exercise, I discovered that this`#if` https://github.com/maxmind/MaxMind-DB-Reader-dotnet/blob/691ba50/MaxMind.Db/GlobalSuppressions.cs#L6  was unnecessary, as `System.Diagnostics.CodeAnalysis` assembly is present in `Microsoft.CodeAnalysis` package(added here: https://github.com/am11/GeoIP2-dotnet/blob/310cc7b5dbf7eaa686f8cb7e9695f63db60d6137/GeoIP2/project.json#L26).

**Comment by oschwald** [comment]
Thanks! I'll take a look tomorrow.

I think we will want to sign the assemblies like we used to. It is a bit of a pain, but it was a very frequent feature request for us. I made a [branch for MaxMind.Db](https://github.com/maxmind/MaxMind-DB-Reader-dotnet/compare/greg/use-strong-name-again?expand=1) that does this and I can do a release so that you can use that.

**Comment by oschwald** [comment]
Related to the signed assembly branch, any idea why [this AppVeyor build](https://ci.appveyor.com/project/oschwald/maxmind-db-reader-dotnet/build/1.0.94/job/3nnrjrr9q2jnr5bl) is failing? It appears to be failing while installing dotnet/cli.

**Comment by am11** [comment]
Hi, I have added back a the signing key. Thanks for the project.json reference! 🎉 

Added the frozen version in CI scripts (to `1.0.0-preview2-002710`) as the latest version file (that `install.[ps1/sh]` downloads from Azure Blob) has been recently added with a commit hash on the first line and version on the second line. This breaks the install script which accepts just one line with version information. I have opened a tracking issue upstream: dotnet/cli#3008.

**Comment by oschwald** [comment]
Thanks for looking into that AppVeyor issue. I'll go ahead and do a signed beta release of MaxMind.Db given that it passes on Travis and locally.

File: GeoIP2/DatabaseReader.cs [hunk 1]
```diff
@@ -423,9 +423,16 @@ private T Execute<T>(string ipStr, IPAddress ipAddress, string type, bool throwO
         {
             if (!Metadata.DatabaseType.Contains(type))
             {
+#if !NETSTANDARD1_4
                 var frame = new StackFrame(2, true);
                 throw new InvalidOperationException(
                     $"A {Metadata.DatabaseType} database cannot be opened with the {frame.GetMethod().Name} method");
+#else
+                // CoreFX is yet to support StackFrame via "System.Diagnostics.StackTrace" contract.
+                // see: https://github.com/dotnet/corefx/issues/179
```
Comment (oschwald) [inline GeoIP2/DatabaseReader.cs]: It this the right issue? It seems to be an unrelated PR.

File: GeoIP2/Http/SyncClient.cs [hunk 2]
```diff
@@ -22,21 +24,29 @@ ProductInfoHeaderValue userAgent
...
+#endif
             request.Headers["Authorization"] = $"Basic {_auth}";
 
             HttpWebResponse response;
             try
             {
+#if !NETSTANDARD1_4
                 response = (HttpWebResponse) request.GetResponse();
+#else
+                response = (HttpWebResponse) request.GetResponseAsync().Result;
```
Comment (oschwald) [inline GeoIP2/Http/SyncClient.cs]: There isn't a synchronous method in netstandard1.4? This change is quite problematic as it doesn't wait for the task to complete, it doesn't handle the exceptions properly, and blocking on async code can lead to deadlocks.

File: GeoIP2.UnitTests/DatabaseReaderTests.cs [hunk 3]
```diff
@@ -9,6 +9,7 @@
 using System.IO;
 using System.Net;
 using System.Reflection;
+using MaxMind.Db.Test;
```
Comment (oschwald) [inline GeoIP2.UnitTests/DatabaseReaderTests.cs]: Why is this necessary?

File: GeoIP2.UnitTests/Program.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Reflection;
+using NUnit.Common;
+using NUnitLite;
+
+#if NETCOREAPP1_0
+using System.IO;
+#endif
+
+namespace MaxMind.Db.Test
```
Comment (oschwald) [inline GeoIP2.UnitTests/Program.cs]: This should probably be moved to the `MaxMind.GeoIP2.UnitTests` namespace.

**Comment by oschwald** [comment]
Overall this looks reasonable, but we'll have to figure out a better solution for the synchronous requests. Where is the best place to find up-to-date information about the status of `HttpWebRequest` in netstandard1.4?

File: GeoIP2/DatabaseReader.cs (hunk 1 above)
Comment (am11) [inline GeoIP2/DatabaseReader.cs]: Copy/paste error, it was supposed to be `1797`. Fixed.

File: GeoIP2.UnitTests/DatabaseReaderTests.cs (hunk 3 above)
Comment (am11) [inline GeoIP2.UnitTests/DatabaseReaderTests.cs]: Fixed the namespace in `Program.cs` and removed this we have `Program.CurrentDirectory` defined there (abstracting out API differences).

File: GeoIP2/Http/SyncClient.cs (hunk 2 above)
Comment (am11) [inline GeoIP2/Http/SyncClient.cs]: Good catch. I have changed it to `GetAwaiter().GetResult()` to avoid blocking the consumer thread. Now it will throw the same exceptions instead of wrapping them in `AggregatedException`. :)

File: GeoIP2.UnitTests/Program.cs (hunk 4 above)
Comment (am11) [inline GeoIP2.UnitTests/Program.cs]: Done.

**Comment by am11** [comment]
I consulted these docs:
* http://dotnet.github.io/api/System.Net.HttpWebRequest.html
* https://github.com/Microsoft/dotnet-apiport/blob/62051f6/docs/RecommendedChanges/System.Net/Use%20HttpClient.SendAsync.md

Seems like they are encouraging everyone to use asynchronous methods when dealing with HTTP. But the situation might change until RTM, as recent CoreFX PRs suggests that they are now sugar coating some missing methods using new underlying infrastructure to mitigate API differences before the RTM.

**Comment by oschwald** [comment]
I wonder if it would make sense to take the same approach for this API and not support the synchronous methods. I am a bit worried that waiting on async methods in synchronous code is going to lead to issues and removing them after a release with them is harder than adding them for .NET Core later.

**Comment by am11** [comment]
Thanks for updating the DB package. I have upgraded the version.

Also, I just found your PR https://github.com/richardszalay/mockhttp/pull/8 and figured out that I was using wrong package in UnitTests project. I have changed that in UnitTests/project.json to your signed package and AppVeyor is succeeding on my fork at the moment: https://ci.appveyor.com/project/am11/geoip2-dotnet/build/1.0.16.

Now there are no more hacks needed in AppVeyor script: yay! 💡 

---

> I wonder if it would make sense to take the same approach for this API and not support the synchronous methods on .NET Core.

I think this is a very good point and reminds me of what Stephen Cleary said in his blog's comment: http://blog.stephencleary.com/2013/11/there-is-no-thread.html

> The method APIs don't lie. They have only synchronous work to do, so they expose a synchronous API.

*(which I have been exercising in ongoing libsass-net wrapper for libsass, where libsass lib only offers sync ops so libsass-net should be transparent about that fact and let consumers `Task.Run` where they want pseudo-async behaviour)*

So in this case, the upstream API (BCL / CoreFX) is providing only asynchrony option, so we would transparently depict only that with .NET Core.

I will `#if` out all the synchronous methods for .NET Core in lib code and tests in sometime. :)

**Comment by oschwald** [comment]
Sounds great. Thanks!

**Comment by am11** [comment]
I have added ef7bd80 for your review. I intend to squash the commit after the CR.

**Comment by oschwald** [comment]
ef7bd80 looks great to me. Thanks!

**Comment by am11** [comment]
Thanks. I have squashed the commits.

**Comment by oschwald** [comment]
Thanks! I merged this via the command line. I am going to do a release as soon as the CI passes for my branch fixing the package name.

**Comment by am11** [comment]
👍 Thanks. I was trying to figure out why Travis CI is complaining with Latest version of CLI (since the install script is fixed upstream). I think we would need to add some dependencies or we can just wait until RC2 is out.

I have gathered that when RC2 is out:

* Remove `"imports": "dnxcore50"` from all three project.jsons in the repo and depend on latest version of JSON.net at that point compatible with NetStandard (which is awaiting RC2 release).
* Remove myget.org from `NuGet.Config` and depend on RC2 packages of CoreFX which will be published on NuGet when RC2 would be announced. This is important because at the moment, anyone wants to depend on GeoIP2 beta needs to have myget.org feed in their config.

**Comment by richardszalay** [comment]
Just a heads up, the latest prerelease mockhttp package supports netstandard1.1 and can be used from a net core app/library.

Be aware that the net45 httpclient  dependency has changed from Microsoft.Net.Http to System.Net.Http on recommendation from MS. The former is still used for net40.

**Comment by oschwald** [comment]
@richardszalay, thanks for letting us know! :+1:
